be/src/exec/pipeline/fragment_context.cpp

```diff
@@ -177,9 +177,17 @@ void FragmentContext::set_stream_load_contexts(const std::vector<StreamLoadConte
     _channel_stream_load = true;
 }
 
+void FragmentContext::release_driver_token_once() {
+    auto* expected = _driver_token_atomic_ptr.load();
+    if (expected != nullptr && _driver_token_atomic_ptr.compare_exchange_strong(expected, nullptr)) {
```

Comment:
If two threads calls `FragmentContext::cancel` concurrently, and then two threads will call `std::unique_ptr::reset` concurrently, which is not thread-safe.

Reply:
Of course, it ensures that `FragmentContext::cancel` is called with non-ok `status` only once, but maybe there is potential bug here.

Reply:
Oh,  just move `_driver_token.reset` to `set_final_status`, since `FragmentContext::cancel` calls `FragmentContext::cancel`, which already apply CAS.

```cpp
void FragmentContext::set_final_status(const Status& status) {
    if (_final_status.load() != nullptr) {
        return;
    }
    Status* old_status = nullptr;
    if (_final_status.compare_exchange_strong(old_status, &_s_status)) {
        _s_status = status;

        _driver_token.reset();
```

Reply:
ok. get your point.

Reply:
if fragmentcontext::cancel should be thread-safe, this unique_ptr may be safer than the following _stream_load_contexts cancel and reset. :)

Reply:
Maybe there is potential bug for `_stream_load_contexts::resize`.

`FragmentContext::finish` also calls `cancel(Status::OK)`. If when a fragment is being finished (call `finish()`) and FE send cancel RPC to this BE, then two threads calls `cancel` concurrently.